src/Mvc/Mvc.Core/src/Formatters/TranscodingReadStream.cs

```diff
@@ -0,0 +1,225 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.Diagnostics;
+using System.IO;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Mvc.Formatters.Json
+{
+    internal sealed class TranscodingReadStream : Stream
+    {
+        internal const int MaxByteBufferSize = 4096;
+        internal const int MaxCharBufferSize = 4 * MaxByteBufferSize;
```

Comment:
Where is this 4x factor coming from?

Reply:
It seemed like a reasonable limit size to for the char buffer. The ctor attempts to use the encoding to determine the max array size, but this is the upper bound for things.

Reply:
If it is primarily for UTF-16 to UTF-8/etc, then 3x is the upper bound.

single UTF-16 character -> 1-3 UTF-8 bytes
surrogate pairs (i.e. 2 UTF-16 characters) -> 4 UTF-8 bytes (i.e. factor is still <= 3x)

Reply:
> If we expect arbitrary encodings, then ignore my comment, since I don't know what the worst case factor would be in that case.

I could change it to 3x. Unicode is the 2nd encoding that we register by default with the formatters, so that makes sense.

Reply:
cc @grabyourpitchforks